dags/synapse-dataset-to-croissant.py

```diff
@@ -165,17 +199,552 @@ def set_up_logging() -> Logger:
+                otel_logger.warning(
+                    f"File {synapse_file.get('file_id')} not found in Synapse.")
+            else:
+                raise ex
+        except SynapseAuthenticationError as ex:
+            if "You are not logged in and do not have access to a requested resource." in str(ex):
+                otel_logger.warning(
+                    f"File {synapse_file.get('file_id')} is not anonymously accessible.")
+            else:
+                raise ex
+    return file_metadata
+
+
+def construct_distribution_section_for_files(files_attached_to_dataset: List[File], **context) -> List[Dict[str, str]]:
+    """
+    Construct the distribution section for the files attached to the dataset. This is
+    used to extract various metadata from the files and create a FileObject for each
+    file.
+
+
+    When we do not have the content_md5 for a file, we need to query Snowflake to get
+    the content_md5 for the file. This is done by querying the synapse_data_warehouse
+    database in Snowflake.
+
+    Arguments:
+        files_attached_to_dataset: The list of files attached to the dataset.
+        context: The context of the DAG run.
+
+    Returns:
+        The distribution section for the files attached to the dataset.
+    """
+    distribution_files = []
+    files_to_find_md5_in_snowflake = {}
+
+    ids_of_files = []
+    id_and_version_pairs = []
+    for file in files_attached_to_dataset:
+        file: File = file
+        if not file.file_handle or not file.file_handle.content_md5:
+            modified_file_id = int(file.id.replace("syn", ""))
+            ids_of_files.append(modified_file_id)
+            id_and_version_pairs.append(modified_file_id)
+            id_and_version_pairs.append(file.version_number)
+            files_to_find_md5_in_snowflake[int(file.id.replace(
+                "syn", ""))] = file.version_number
+
+    file_md5s = None
+    if ids_of_files:
+        snow_hook = SnowflakeHook(
+            context["params"]["snowflake_developer_service_conn"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+
+        query = f"""
+        WITH version_data AS (
+            SELECT
+                nl.id,
+                flattened.value:versionNumber::int AS versionNumber,
+                flattened.value:contentMd5::string AS contentMd5
+            FROM synapse_data_warehouse.synapse.node_latest AS nl,
+            LATERAL FLATTEN(input => nl.VERSION_HISTORY) AS flattened
+            WHERE nl.id IN ({', '.join(['%s'] * len(ids_of_files))})
+            AND nl.is_public = TRUE
+        )
+        SELECT
+            vd.id,
+            vd.versionNumber,
+            vd.contentMd5
+        FROM version_data AS vd
+        WHERE (vd.id, vd.versionNumber) IN ({', '.join(['(%s, %s)'] * (len(id_and_version_pairs)//2))});
+        """
+        try:
+            cs.execute(
+                query,
+                (ids_of_files + id_and_version_pairs),
```

Comment:
I wonder if snowflake has limits to the number of `value, value2` in the query line

Reply:
I couldn't find any limit on the number of placeholders that may be used. I did find limits that the query length can be up to 1 MB, which, I don't think we would hit even if we were using a dataset that had 30k entities on it (It'd be 90k placeholder strings). It would only be around 500 KB